### PR TITLE
Move mailserver logic/gaps/ranges to status-go

### PR DESCRIPTION
**Pull request by cammellos**
This commit moves most of the mailserver logic to status-go.
    
- Filters are now removed and not passed to the client anymore
- Ranges have been removed
- Gaps are now messages with a different content type
- Upsert/Save chat has been removed and instead we have more specific
    endpoints such as CreatePublicChat/CreateOneToOneChat/CreateProfileChat
- Creation of timeline/profile chat has been moved to status-go

### CODE REVIEW

**Comment by churik** [comment]
@cammellos 
based on e2e:
### ISSUE 1: Clear history doesn't remove all history from chat and it reappers after relogin

**Steps:**
1) restore multiaccount from seed phrase
2) join any public chat
3) "Clear history" and go back to home view
4) religion and check chat

**Expected result:**
3) history is cleared, no message in chat view, preview message is empty on home view
4) cleared history doesn't  repappear
**Actual result:**
3) several messages are cleared, previous history is still shown in  chat view, , preview message is empty on home view
4) cleared history   repappears
**OS:** IOS, Android

https://user-images.githubusercontent.com/4557972/117942392-4c3cd780-b30b-11eb-8b71-080a24757522.mp4

### ISSUE 2: Endless loading spinner when creating group chat or receiving invite to group chat
IMAGE 2021-05-12 10:24:45

**OS:** IOS, Android

### ISSUE 3: No redirect when declining invite to group chat or leaving group chat via group info
**Steps:**
1) User A: restore multiaccount from seed phrase
2) User A: create group chat with User B
3) User B: Decline invitation

**Expected result:**
will be redirected to home view
**Actual result:**
reminding on chat view, visually nothing happens
**OS:** IOS, Android

### ISSUE 4: Reconnecting to another mailserver when relogin if "Automatic selection" is off and connected to custom mailserver

**Steps:**
1) restore multiaccount from seed phrase
2) go to Profile > Sync settings > Status nodes
3) switch off "Automatic selection"
4) add custom mailserver and connect to it (I've added non-working mailserver `enode://69f72baa7f1722d111a8c9c68c39a31530e9d567695f6108f31ccb6cd8f0adff5991e7fdca8fa770e75bc8a511a87d25690cbc80e008175f50c157d6f6788d58:status-offline-inbox@206.189.250.16:553`)
5) religion and check Status nodes

**Expected result:** still connected to custom mailserver

**Actual result:** connected to `mail-03.do-ams-3` (located in Spain), automatic selection is off

**OS:** IOS, Android

**Comment by cammellos** [comment]
@churik thank you! looking into it

**Comment by Serhy** [comment]
With latest builds `2b409b4` (they set to 'staging' by default) there are few issues: 
1) Offline inbox message history is not fetching chats. To reproduce, - just send in any chat where recipient is offline, text message(s). When recipient logs in to account - no messages are fetched. In 1-1/Group chat on message Sender side it's visible that recipient not read the message (one checkmark displayed below sent. messages). Same applies to 'Timeline' (Status tab) - no status updates are fetched.

2) In Group chat after re-login there is endless 'Loading...' spinner at the top of chat.
Screenshot 2021-05-17 at 12 21 12

3) Same as 2nd item above but with different steps for public chat:
- UserA and UserB join the same public chat
- UserB stays on Chats main view 
- UserA sends text message to public chat
- UserB sees unread indicator
- UserB re-login to account
- UserB opens public chat
**Actual result:**  Endless `Loading...` spinner instead of 'Fetch more messages'

File: src/status_im/chat/db.cljs [hunk 1]
```diff
@@ -48,89 +46,36 @@
...
-          (update acc :messages conj {:type       :gap
-                                      :value      (str :first-gap)
-                                      :first-gap? true})
-          acc)))
-     ([{:keys [messages previous-message gaps]} message]
-      (let [{:keys [gaps-number gap]}
-            (check-gap gaps previous-message message)
-            add-gap?      (pos? gaps-number)]
-        {:messages           (cond-> messages
+(def map->sorted-seq
```
Comment (flexsurfer) [inline src/status_im/chat/db.cljs]: is not used?can be removed

File: src/status_im/chat/db.cljs [hunk 2]
```diff
@@ -48,89 +46,36 @@
...
+  (comp (partial map second) (partial sort-by first)))
 
-                               add-gap?
-                               (add-gap gap)
+(defn collapse-gaps [messages chat-id synced-from]
```
Comment (flexsurfer) [inline src/status_im/chat/db.cljs]: could you please add the doc to the function, why do we need it and what its doing, thanks

File: src/status_im/chat/db.cljs [hunk 3]
```diff
@@ -48,89 +46,36 @@
...
+                                 ;; it's a gap
+                                  :else
+                                  (conj acc (assoc message :gap-ids #{message-id})))))
+                            []
+                            messages)]
+    ;; If it's a gap or the chat is still syncing, do nothing
+    (if (or (nil? synced-from)
+            (:gap-ids (peek messages-with-gaps)))
+      messages-with-gaps
+      (conj messages-with-gaps {:message-id "0x123"
```
Comment (flexsurfer) [inline src/status_im/chat/db.cljs]: what is this map and why we conj it ?

File: src/status_im/chat/models.cljs [hunk 4]
```diff
@@ -209,24 +137,39 @@
...
+                                    :deleted-at-clock-value    last-message-clock-value}))}))
+
+(fx/defn clear-history-handler
+  "Clears history of the particular chat"
+  {:events [:chat.ui/clear-history]}
+  [{:keys [db] :as cofx} chat-id remove-chat?]
+  (fx/merge cofx
+            {:db db
+             ::json-rpc/call [{:method "wakuext_clearHistory"
+                               :params [{:ID chat-id}]
```
Comment (flexsurfer) [inline src/status_im/chat/models.cljs]: why ID and not id ?

File: src/status_im/data_store/settings.cljs [hunk 5]
```diff
@@ -1,11 +1,15 @@
 (ns status-im.data-store.settings
-  (:require [status-im.ethereum.eip55 :as eip55]))
+  (:require
```
Comment (flexsurfer) [inline src/status_im/data_store/settings.cljs]: just wondering how's this related to this PR?

File: src/status_im/ethereum/json_rpc.cljs [hunk 6]
```diff
@@ -36,6 +36,13 @@
    "waku_getSymKey" {}
    "waku_markTrustedPeer" {}
    "wakuext_post" {}
+   "wakuext_requestAllHistoricMessages" {}
+   "wakuext_fillGaps" {}
+   "wakuxet_syncChatFromSyncedFrom" {}
```
Comment (flexsurfer) [inline src/status_im/ethereum/json_rpc.cljs]: wakuxet_syncChatFromSyncedFrom duplicate with typo?

File: src/status_im/subs.cljs [hunk 7]
```diff
@@ -946,18 +934,21 @@
...
+    (re-frame/subscribe [:chats/message-list chat-id])
     (re-frame/subscribe [:chats/chat-messages chat-id])
-    (re-frame/subscribe [:chats/messages-gaps chat-id])
-    (re-frame/subscribe [:chats/range chat-id])
-    (re-frame/subscribe [:chats/all-loaded? chat-id])
-    (re-frame/subscribe [:chats/public? chat-id])])
- (fn [[message-list messages messages-gaps range all-loaded? public?]]
+    (re-frame/subscribe [:chats/loading-messages? chat-id])
+    (re-frame/subscribe [:chats/synced-from chat-id])])
+ (fn [[chat-id message-list messages loading-messages? synced-from]]
```
Comment (flexsurfer) [inline src/status_im/subs.cljs]: `query-vec` available here as a second param `(fn [[message-list messages loading-messages? synced-from] [_ chat-id]]`

File: src/status_im/ui/screens/chat/message/message.cljs [hunk 8]
```diff
@@ -316,6 +317,11 @@
   [message]
   [message.command/command-content message-content-wrapper message])
 
+(defmethod ->message constants/content-type-gap
+  [message]
+  (println "GAP" message)
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/message/message.cljs]: leftover?

File: src/status_im/ui/screens/chat/views.cljs [hunk 9]
```diff
@@ -289,6 +290,7 @@
   (let [{:keys [group-chat chat-id public?]} chat
         messages @(re-frame/subscribe [:chats/chat-messages-stream chat-id])
         current-public-key @(re-frame/subscribe [:multiaccount/public-key])]
+    (println "CHAT" chat)
```
Comment (flexsurfer) [inline src/status_im/ui/screens/chat/views.cljs]: leftover?

File: src/status_im/utils/prices.cljs [hunk 10]
```diff
@@ -39,10 +39,10 @@
                                      :last-day (:OPEN24HOUR entry)}}))}))))
 
 (defn get-prices [from to mainnet? on-success on-error]
-  (log/debug "[prices] get-prices"
-             "from" from
-             "to" to
-             "mainnet?" mainnet?)
+  (log/info "[prices] get-prices"
```
Comment (flexsurfer) [inline src/status_im/utils/prices.cljs]: ?

File: src/status_im/chat/models.cljs [hunk 11]
```diff
@@ -277,65 +213,91 @@
...
 
 (defn my-profile-chat-topic [db]
   (profile-chat-topic (get-in db [:multiaccount :public-key])))
 
+(fx/defn handle-public-chat-created
+  {:events [::public-chat-created]}
+  [{:keys [db] :as cofx} chat-id {:keys [dont-navigate?]} response]
+  (let [chat (chats-store/<-rpc (first (:chats response)))]
+    {:db (assoc-in db [:chats chat-id] chat)
+     :dispatch (when-not dont-navigate?
```
Comment (flexsurfer) [inline src/status_im/chat/models.cljs]: i believe there will be an error for `:dispatch nil`

**Review by flexsurfer — APPROVED** [review]
WOW, just amazing job!!

Alt Text

File: src/status_im/chat/db.cljs (hunk 3 above)
Comment (cammellos) [reply src/status_im/chat/db.cljs]: It's the last gap, which is not saved in the database, I will separate into a function and add a comment

**Comment by cammellos** [comment]
@flexsurfer addressed feedback, thank you!

**Comment by Serhy** [comment]
Tested on Android11 and iOS 14, and no issues found (apart two below which agreed to fix separately). 
- Messaging in all chat types
- Clear history/delete chats
- Offline history
- Timeline and mobile network connection
- Blocking users etc

Also tested release 1.13 on staging fleet which looks good as well.
Leftover issue to fix separately:
- After pairing accounts public chat synced fine but no messages are fetched automatically in it. Need to re-login or Fetch messages.
- Gaps in Group chats at the beginning of chat history and Fetch more messages at the top of chat which does nothing when tap
@cammellos please don't forget to revert 'staging' as default fleet back to prod. Thanks a lot for that big useful changes!

**Pull request merged**